### PR TITLE
Add Ask User Tool to AI Assistant

### PR DESCRIPTION
**Pull request by bryantgillespie**
## Scope

image

https://github.com/user-attachments/assets/610af5a6-1ec0-41d5-a18b-2f302155ffdd

Adds an "Ask User" tool to the AI Assistant so it can pause and ask you questions (single choice, multi-select, or free text) instead of guessing what you meant. Chat input swaps to a dedicated component when a question is pending, and answers get summarized in message history.

### What's changed:

- `use-ask-user-tool` composable handling the full lifecycle — registration, answers, cancellation, edge cases
- Conversation UI integration with `AiAskUser` component and summary view
- English translations for new UI labels

## Potential Risks / Drawbacks

- LLMs love overusing this pattern. Added prompting guardrails but worth watching.

## Tested Scenarios

- Single choice, multi-select, free text
- Cancellation mid-question
- Edge cases (empty answers, rapid submissions)

## Review Notes / Questions
- This is a table stakes feature for most chatbot interactions these days and is definitely a UX improvement on what we have now.
- Didn't have a previous mocked design for this but Design team has already approved.

## Checklist

- [x] Added or updated tests

---

###  Copilot Summary
This pull request introduces the new "Ask User" tool to the AI Assistant, enabling AI workflows to pause and prompt the user for input via configurable questions. It includes the tool's implementation, integration into the conversation UI, and comprehensive tests. The changes also add summary and localization support for the new tool.

**AI Assistant: Ask User Tool Integration**

*Feature Implementation & Integration:*
- Added the "Ask User" tool, allowing AI to ask users one or more questions and wait for their answers, with support for options, free text, and multi-select. (`app/src/ai/composables/use-ask-user-tool.ts`, `.changeset/wild-showers-push.md`) [[1]](diffhunk://#diff-eee722700c71ba3ec6e243deb8be27cf10cf33d7cdb7532975e2be43a4e27620R1-R91) [[2]](diffhunk://#diff-fb4d7356bcc30b87a6c9b3d2bc0f96880d2090a85c2a98fc842a7a41330a6ac6R1-R5)
- Integrated the tool into the AI conversation UI: when a question is pending, the input switches to an `AiAskUser` component; responses are summarized in message history. (`app/src/ai/components/ai-conversation.vue`, `app/src/ai/components/parts/ai-ask-user-summary.vue`, `app/src/ai/components/parts/ai-message-tool.vue`) [[1]](diffhunk://#diff-2b3a7ccbe47e868097cce7f2eba46e4ffaf3761267ca2bea7319889858e7af16R4-R9) [[2]](diffhunk://#diff-2b3a7ccbe47e868097cce7f2eba46e4ffaf3761267ca2bea7319889858e7af16R19-R20) [[3]](diffhunk://#diff-2b3a7ccbe47e868097cce7f2eba46e4ffaf3761267ca2bea7319889858e7af16L117-R122) [[4]](diffhunk://#diff-4c566081501af228ef0214d8e3d41587366c271859a441f0eea1f97a7fadd392R1-R62) [[5]](diffhunk://#diff-ba9a0fd78506ab6246963e5ef5b4630f6ac7061fbf2a5ad3f84af54cf6f611f5R5) [[6]](diffhunk://#diff-ba9a0fd78506ab6246963e5ef5b4630f6ac7061fbf2a5ad3f84af54cf6f611f5R42-R58)

*Testing:*
- Added comprehensive tests for the Ask User tool's lifecycle, including registration, answer submission, cancellation, and edge cases. (`app/src/ai/composables/use-ask-user-tool.test.ts`)

*Localization & Display:*
- Added English translations for Ask User tool UI labels and summary display. (`app/src/lang/translations/en-US.yaml`) [[1]](diffhunk://#diff-df42ecafe708c5d98ddbebfb3a7238207aab0c3c3c93920d616221136c08ad6aR1846-R1851) [[2]](diffhunk://#diff-df42ecafe708c5d98ddbebfb3a7238207aab0c3c3c93920d616221136c08ad6aR1864)

### CODE REVIEW

**Comment by bryantgillespie** [comment]
@claude Can you help me out with a review on this guy?

**Comment by bryantgillespie** [comment]
@claude did we address all your concerns here my guy?

**Comment by bryantgillespie** [comment]
> Two minor issues remain:
> 
> Zod validation - should prevent configurations with no input methods (recommended fix)
> Component unmount safety - very low risk edge case with question transitions
> The core functionality is solid and the major concerns (ResizeObserver leak, accessibility) have been addressed. The remaining issues are edge cases that are unlikely to cause problems in normal usage, but the Zod validation would be a good safeguard to add.
> 
> Recommendation: Consider adding the Zod refinement for completeness, but the PR is otherwise ready to merge.

@claude zod doesn't work well with converting to JSON Schemas and using refine. So let's skip that part. I don't think the remaining concern about unmounting during transition is a big problem either.

Can you resolve your comments?

File: app/src/ai/components/parts/ai-ask-user.vue [hunk 1]
```diff
@@ -0,0 +1,615 @@
...
+		resizeObserver = new ResizeObserver(updateFades);
+		resizeObserver.observe(tabsContainerRef.value);
+	}
+});
+
+onUnmounted(() => {
+	resizeObserver?.disconnect();
+	// Safety net for non-standard unmount paths (e.g. parent route change)
+	cancelPending();
+});
```
Comment (alvarosabu) [inline app/src/ai/components/parts/ai-ask-user.vue:234]: Suggestion:

@bryantgillespie you can directly use [`useResizeObserver`](https://vueuse.org/core/useResizeObserver/) from vueuse so you dont need to manually create/disconnect on the lifecycles.

```ts
  import { useResizeObserver } from '@vueuse/core';

  useResizeObserver(tabsContainerRef, () => {
    updateFades();
  });
```

File: app/src/ai/components/parts/ai-ask-user.vue [hunk 2]
```diff
@@ -0,0 +1,615 @@
...
+		case 'Escape':
+			event.preventDefault();
+
+			if (!isLastQuestion.value) {
+				goToQuestion(activeQuestionIndex.value + 1);
+			}
+
+			break;
+	}
+}
```
Comment (alvarosabu) [inline app/src/ai/components/parts/ai-ask-user.vue]: Suggestion: 

This method with switch can be replaced with [`onKeyStroke` composable](https://vueuse.org/core/onKeyStroke/#onkeystroke) from vueuse 

```typescript
import { onKeyStroke } from '@vueuse/core';

// Arrow key navigation
onKeyStroke('ArrowLeft', (e) => {
	e.preventDefault();
	goToQuestion(activeQuestionIndex.value - 1);
});

onKeyStroke('ArrowRight', (e) => {
	e.preventDefault();
	goToQuestion(activeQuestionIndex.value + 1);
});

// Enter key handling
onKeyStroke('Enter', (e) => {
	if (!currentQuestion.value) return;
	const q = currentQuestion.value;
	if (e.shiftKey || textInputActive.value[q.id]) return;

	e.preventDefault();
	if (isLastQuestion.value) {
		handleSubmit();
	} else {
		goToQuestion(activeQuestionIndex.value + 1);
	}
});

// Escape key handling
onKeyStroke('Escape', (e) => {
	e.preventDefault();
	if (!isLastQuestion.value) {
		goToQuestion(activeQuestionIndex.value + 1);
	}
});

// Number keys 1-4 for option selection
onKeyStroke(['1', '2', '3', '4'], (e) => {
	if (!currentQuestion.value) return;
	const q = currentQuestion.value;
	const num = parseInt(e.key);

	if (num >= 1 && num <= (q.options?.length ?? 0) && q.options) {
		e.preventDefault();
		selectOption(q.id, q.options[num - 1]!.label, q.multi_select);
	}
});
```

File: app/src/ai/components/parts/ai-ask-user.vue [hunk 3]
```diff
@@ -0,0 +1,615 @@
...
+</script>
+
+<template>
+	<div
+		ref="rootEl"
+		class="ai-ask-user"
+		role="region"
+		:aria-label="t('ai.ask_user')"
+		tabindex="0"
+		@keydown="handleKeydown"
```
Comment (alvarosabu) [inline app/src/ai/components/parts/ai-ask-user.vue]: If you use the composable I mentioned above you can remove this from here. VueUse optimizes event listeners

File: app/src/ai/components/parts/ai-ask-user.vue [hunk 4]
```diff
@@ -0,0 +1,615 @@
...
+	if (!currentQuestion.value) return;
+
+	const q = currentQuestion.value;
+
+	// Number keys 1-4 to select options
+	const num = parseInt(event.key);
+
+	if (num >= 1 && num <= (q.options?.length ?? 0) && q.options) {
+		event.preventDefault();
+		selectOption(q.id, q.options[num - 1]!.label, q.multi_select);
```
Comment (alvarosabu) [inline app/src/ai/components/parts/ai-ask-user.vue]: `q.options[num - 1]!.label` uses non-null assertion. Since options is optional in the
  schema and the length check uses `??`, this could be safer:

```ts
  const option = q.options?.[num - 1];
  if (option) selectOption(q.id, option.label, q.multi_select);
```

File: app/src/ai/components/parts/ai-ask-user.vue [hunk 5]
```diff
@@ -0,0 +1,615 @@
...
+	const newHeight = wrapper.offsetHeight;
+	wrapper.style.height = `${savedHeight}px`;
+
+	requestAnimationFrame(() => {
+		wrapper.style.height = `${newHeight}px`;
+	});
+}
+
+function onAfterEnter() {
+	const wrapper = questionWrapperRef.value;
```
Comment (alvarosabu) [inline app/src/ai/components/parts/ai-ask-user.vue:75]: Blocker:

`requestAnimationFrame` is never canceled. If the component unmounts mod-animation, it  will fire on a stale ref. You can add a `cancelAnimationFrame` in `onUnmounted`

File: app/src/ai/components/parts/ai-ask-user.vue [hunk 6]
```diff
@@ -0,0 +1,615 @@
...
+	return isAnswered(currentQuestion.value.id);
+});
+
+const isLastQuestion = computed(() => activeQuestionIndex.value === questions.value.length - 1);
+
+const slideDirection = ref<'forward' | 'backward'>('forward');
+
+const rootEl = ref<HTMLElement | null>(null);
+const questionWrapperRef = useTemplateRef<HTMLElement>('question-wrapper');
+let savedHeight = 0;
```
Comment (alvarosabu) [inline app/src/ai/components/parts/ai-ask-user.vue]: This should be a ref not a let.

```ts
const savedHeight = ref(0)
```

You will need to then update all occurrences with `savedHeight.value`

File: app/src/ai/components/parts/ai-ask-user.vue [hunk 7]
```diff
@@ -0,0 +1,615 @@
...
+
+function onBeforeLeave() {
+	const wrapper = questionWrapperRef.value;
+	if (wrapper) savedHeight = wrapper.offsetHeight;
```
Comment (alvarosabu) [inline app/src/ai/components/parts/ai-ask-user.vue]:
```suggestion
	if (wrapper) savedHeight.value = wrapper.offsetHeight;
```

File: app/src/ai/components/parts/ai-ask-user-summary.vue [hunk 8]
```diff
@@ -0,0 +1,62 @@
...
+import type { DynamicToolUIPart } from 'ai';
+import { computed } from 'vue';
+import type { AskUserAnswers, AskUserInput } from '../../composables/use-ask-user-tool';
+
+const props = defineProps<{
+	part: DynamicToolUIPart;
+}>();
+
+const questions = computed(() => (props.part.input as AskUserInput | undefined)?.questions ?? []);
+const output = computed(() => (props.part.output as AskUserAnswers) ?? {});
```
Comment (alvarosabu) [inline app/src/ai/components/parts/ai-ask-user-summary.vue]: These raw as casts bypass type safety. If the `DynamicToolUIPart` type can be narrowed (e.g., by checking
  part.type === 'tool-ask_user'), a type guard would be safer.

**Review by alvarosabu — CHANGES_REQUESTED** [review]
Is looking good @bryantgillespie nice work here, I just had time to check it codewise, I will play with it tomorrow.

If you have any question regarding my comments or you need help form my side please ping me.

**Comment by alvarosabu** [comment]
The post category doesn't exist in the schema. Maybe we should add something to restrict AI to ask only existing fields.

Screenshot 2026-02-17 at 11 08 05

**Comment by alvarosabu** [comment]
Author field is a o2m relationship, from UX perspective I would expect a dropdown with the options rather than manually enter the name or id.

Screenshot 2026-02-17 at 11 13 52

**Comment by alvarosabu** [comment]
After planning before creating, another choice is asked but this time in the form of text instead of the card with options. Can be this improved or we have a limitation?

Screenshot 2026-02-17 at 11 16 15

**Comment by alvarosabu** [comment]
After doing some operation, takes a long time between one exectution and the feedback/another-execution, it gives the feeling is completly unresponsive, maybe we should add a "Thinking" or any loading progression visual indicator

Screenshot 2026-02-17 at 11 18 01

Something like 
Screenshot 2026-02-17 at 11 19 58

**Comment by bryantgillespie** [comment]
> a long time between one exectution and the feedback/another-execution, it gives the feeling is completly unresponsive, maybe we should add a "Thinking" or any loading progression visual indicator

yeah for sure -- this has been one we need to add for a minute now. we have the little indicator initially but we just need to show in the middle of tool calling while waiting. will do that in a separate PR.

File: app/src/ai/components/parts/ai-ask-user.vue (hunk 1 above)
Comment (alvarosabu) [reply app/src/ai/components/parts/ai-ask-user.vue:234]: I addressed this here https://github.com/directus/directus/pull/26681/commits/c4b2ed551d717c9f53e3e9f03034a170d2c04508 https://github.com/directus/directus/pull/26633

File: app/src/ai/components/parts/ai-ask-user.vue (hunk 2 above)
Comment (alvarosabu) [reply app/src/ai/components/parts/ai-ask-user.vue]: Done here https://github.com/directus/directus/pull/26681/commits/52bcd331c79e39561df77b4e0084cb17813efe87 https://github.com/directus/directus/pull/26633

File: app/src/ai/components/parts/ai-ask-user.vue (hunk 3 above)
Comment (alvarosabu) [reply app/src/ai/components/parts/ai-ask-user.vue]: Done here https://github.com/directus/directus/commit/52bcd331c79e39561df77b4e0084cb17813efe87 https://github.com/directus/directus/pull/26633

File: app/src/ai/components/parts/ai-ask-user.vue (hunk 6 above)
Comment (alvarosabu) [reply app/src/ai/components/parts/ai-ask-user.vue]: Done here https://github.com/directus/directus/commit/52bcd331c79e39561df77b4e0084cb17813efe87 https://github.com/directus/directus/pull/26633

File: app/src/ai/components/parts/ai-ask-user.vue (hunk 7 above)
Comment (alvarosabu) [reply app/src/ai/components/parts/ai-ask-user.vue]: Done here https://github.com/directus/directus/commit/52bcd331c79e39561df77b4e0084cb17813efe87 https://github.com/directus/directus/pull/26633

File: app/src/ai/components/parts/ai-ask-user.vue (hunk 5 above)
Comment (alvarosabu) [reply app/src/ai/components/parts/ai-ask-user.vue:75]: Resolved here https://github.com/directus/directus/pull/26687/commits/0039d13e7f03866e50d014803c15e70190f0bb80

**Review by alvarosabu — APPROVED** [review]
LGTM @bryantgillespie 🚢

**Review by AlexGaillard — APPROVED** [review]
:cool::beans:

**Review by AlexGaillard — CHANGES_REQUESTED** [review]
Image

Scratch that approval. You are not able to type the number "1" into the input box

**Review by AlexGaillard — APPROVED** [review]
